File: python/panel_reverberation.py
from cycle import Cycle


class PanelReverberation:
    def __init__(self, panel, interaction, interaction_config):
        self.panel = panel
        self.interaction = interaction
        self.interaction_config = interaction_config
        self.scale = 0.0
        oneShot = not self.is_source_interaction()
        self.cycle = Cycle(self.interaction_config.initial_trigger_panel_animation_loop_duration_ticks,
                           lambda value: self.on_up_cycle(value), lambda value: self.on_down_cycle(value), oneShot)
        self.current_value = 0.0
        self.distance_from_trigger = self.interaction.get_distance_from_trigger_to_panel(self.panel)

    def to_dict(self):
        return {
            "panel": self.panel.to_dict(),
            "interactionSourcePanel": self.interaction.source_panel.to_dict(),
            "distanceFromTrigger": self.distance_from_trigger,
            "scale": self.scale,
            "currentValue": self.current_value
        }

    def _calculate_scale(self):
        if self.panel == self.interaction.source_panel:
            return 1
        else:
            maxDistance = self.interaction.current_reverberating_distance
            return (maxDistance - self.distance_from_trigger) / maxDistance

    def is_source_interaction(self):
        return self.panel.index == self.interaction.source_panel.index

    def start(self):
        self.scale = self._calculate_scale()
        self.cycle.restart()
        # console.log("Started PanelReverberation", self)

    def stop(self):
        self.cycle.stop()
        # console.log("Stopped PanelReverberation", self)

    def on_up_cycle(self, value):
        # on our way up
        self._set_value(value)
        if not self.interaction.source_panel.interaction_active and not self.cycle.is_at_zero_point():
            # Fade out from where self animation loop is at by jumping the cycle to the down-side.
            self.cycle.jumpToDownCycle()

    def on_down_cycle(self, value):
        # on our way down
        self._set_value(value)

    def _set_value(self, value):
        self.current_value = value * self.scale

    def update(self):
        self.cycle.next()
        if self.is_source_interaction() and self.panel.interaction_active and self._is_animation_loop_done():
            self.cycle.restart(self.interaction_config.get_trigger_panel_animation_loop_duration_ticks())

    def _is_animation_loop_done(self):
        return self.cycle.is_done() or (self.cycle.iterations > 0 and self.cycle.is_at_zero_point())

    def is_done(self):
        if self.is_source_interaction() and self.panel.interaction_active:
            return False

        if not self.cycle.clock.running:
            return True

        # Otherwise, self is waiting for the animation loop to be done.
        return self._is_animation_loop_done()

File: arduino/src/PanelReverberation.cpp
//
// Created by David Clay on 6/15/23.
//

#include "PanelReverberation.h"

PanelReverberation::PanelReverberation(Panel* panel,
                                       Panel* sourcePanel,
                                       int distanceFromTrigger,
                                       ReverberationDistanceSource* reverberationDistanceSource,
                                       Config* config) {
    this->panel = panel;
    this->sourcePanel = sourcePanel;
    this->config = config;
    this->reverberationDistanceSource = reverberationDistanceSource;
    this->isSourceInteraction = panel->index == this->sourcePanel->index;
    this->scale = 0;
    bool oneShot = !this->isSourceInteraction;
    this->cycle = new Cycle(config->initial_trigger_panel_animation_loop_duration_ticks,
                            oneShot,
                            CYCLE_TYPE_UP_AND_DOWN,
                            this);
    this->currentValue = 0;
    this->distanceFromTrigger = distanceFromTrigger;
}

void PanelReverberation::value(float value, CycleDirection direction) {
    this->setValue(value);
    if (direction == CYCLE_DIRECTION_UP) {
        if (!this->sourcePanel->active && !this->cycle->isAtZeroPoint()) {
            this->cycle->jumpToDownCycle();
        }
    }
}

void PanelReverberation::setValue(float value) {
    this->currentValue = value * this->scale;
}

void PanelReverberation::start() {
    this->scale = this->calculateScale();
    this->cycle->restart();
}

void PanelReverberation::stop() const {
    this->cycle->stop();
}

void PanelReverberation::update() const {
    this->cycle->next();
    if (this->isSourceInteraction && this->panel->active && this->isAnimationLoopDone()) {
        int newDurationTicks = this->config->getTriggerPanelAnimationLoopDurationTicks();
        this->cycle->restart(newDurationTicks);
    }
}

bool PanelReverberation::isDone() const {
    if (this->isSourceInteraction && this->panel->active) {
        return false;
    }

    if (!this->cycle->clock->running) {
        return true;
    }

    return this->isAnimationLoopDone();
}

float PanelReverberation::calculateScale() const {
    if (this->isSourceInteraction) {
        return 1;
    } else {
        int maxDistance = this->reverberationDistanceSource->getReverberationDistance();
        return (float)(maxDistance - this->distanceFromTrigger) / (float)maxDistance;
    }
}

bool PanelReverberation::isAnimationLoopDone() const {
    return this->cycle->isDone() || (this->cycle->iterations > 0 && this->cycle->isAtZeroPoint());
}
